### PR TITLE
Allow chain selection for different blocks with same `SelectView`

### PR DESCRIPTION
**Pull request by amesgen**
# Description

Clarify under which circumstances we can compare different blocks (see the first commit message), which enables boilerplate reduction in `hardForkChainSel` for the Cardano eras.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Protocol/ChainSel.hs [hunk 1]
```diff
@@ -39,18 +39,11 @@ data AcrossEraSelection :: Type -> Type -> Type where
...
-  -- NOTE: We require that the eras have the same /protocol/, not merely the
-  -- same 'SelectView', because if we have two eras with different protocols
-  -- that happen to use the same 'SelectView' but a different way to compare
-  -- chains, it's not clear how to do cross-era selection.
-  SelectSameProtocol ::
-       BlockProtocol x ~ BlockProtocol y
+  -- | Two eras using the same 'SelectView'. In this case, we can just compare
+  -- chains even across eras, as the chain ordering is fully captured by
+  -- 'SelectView' and its 'Ord' instance.
+  SelectSameSelectView ::
```
Comment (amesgen) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Protocol/ChainSel.hs:45]: Any better naming suggestions?

Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Protocol/ChainSel.hs:45]: `CompareSameSelectView`?

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Protocol/ChainSel.hs [hunk 2]
```diff
@@ -76,9 +69,9 @@ acrossEras ::
   -> Ordering
 acrossEras (WithBlockNo bnoL (WrapSelectView l))
            (WithBlockNo bnoR (WrapSelectView r)) = \case
-    CompareBlockNo     -> compare bnoL bnoR
-    CustomChainSel f   -> f l r
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Protocol/ChainSel.hs:80]: I would guess `CustomChainSel` doesn't occur in the codebase now.

Should we remove it, and only add it back if/when the future day arises where we actually need it?

**Review by nfrisby — APPROVED** [review]
Approved 👍 

I left a couple minor comments.

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Protocol/ChainSel.hs (hunk 2 above)
Comment (amesgen) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Protocol/ChainSel.hs:80]: Sounds good, adding back `CustomChainSel` will probably the easiest part when we have pairs of blocks/eras that can neither use `CompareBlockNo` nor `CompareSameSelectView`.

**Comment by amesgen** [comment]
bors merge

**Pull request merged**